src/lib.rs

```diff
@@ -382,20 +403,27 @@ pub fn optimize(input: &InFile, output: &OutFile, opts: &Options) -> PngResult<(
                     err
                 ))
             })?;
-            if opts.preserve_attrs {
-                if let Some(input_path) = input.path() {
-                    copy_permissions(input_path, &out_file);
-                }
+            if let Some(metadata_input) = &opt_metadata_preserved {
+                copy_permissions(&metadata_input, &out_file);
             }
 
             let mut buffer = BufWriter::new(out_file);
-            buffer.write_all(&optimized_output).map_err(|e| {
-                PngError::new(&format!(
-                    "Unable to write to {}: {}",
-                    output_path.display(),
-                    e
-                ))
-            })?;
+            buffer
+                .write_all(&optimized_output)
+                // flush BufWriter so IO errors don't get swallowed silently on close() by drop!
+                .and_then(|()| buffer.flush())
+                .map_err(|e| {
+                    PngError::new(&format!(
+                        "Unable to write to {}: {}",
+                        output_path.display(),
```

Comment:
I think you don't need to explicitly specify `display()` here--by default, when passing an argument to `format` as a `{}` parameter, it will call the Display impl.